python/instrumentation/openinference-instrumentation-haystack/src/openinference/instrumentation/haystack/_wrappers.py

```diff
@@ -0,0 +1,207 @@
+from abc import ABC
+from typing import Any, Callable, Iterator, List, Mapping, Tuple
+from enum import Enum
+from openinference.semconv.trace import (
+    OpenInferenceSpanKindValues,
+    SpanAttributes,
+    EmbeddingAttributes,
+    OpenInferenceMimeTypeValues,
+    DocumentAttributes
+)
+from opentelemetry import trace as trace_api
+import opentelemetry.context as context_api
+from opentelemetry.util.types import AttributeValue
+from openinference.instrumentation import get_attributes_from_context, safe_json_dumps
+
+
+def _flatten(mapping: Mapping[str, Any]) -> Iterator[Tuple[str, AttributeValue]]:
+    for key, value in mapping.items():
+        if value is None:
+            continue
+        if isinstance(value, Mapping):
+            for sub_key, sub_value in _flatten(value):
+                yield f"{key}.{sub_key}", sub_value
+        elif isinstance(value, List) and any(isinstance(item, Mapping) for item in value):
+            for index, sub_mapping in enumerate(value):
+                for sub_key, sub_value in _flatten(sub_mapping):
+                    yield f"{key}.{index}.{sub_key}", sub_value
+        else:
+            if isinstance(value, Enum):
+                value = value.value
+            yield key, value
+
+class _WithTracer(ABC):
+    """
+    Base class for wrappers that need a tracer. Acts as a trait for the wrappers
+    """
+
+    def __init__(self, tracer: trace_api.Tracer, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+        self._tracer = tracer
+
+
+class _ComponentWrapper(_WithTracer):
+    """
+    Wrapper for the pipeline processing
+    Captures all calls to the pipeline
+    """
+    def __call__(
+            self,
+            wrapped: Callable[..., Any],
+            instance: Any,
+            args: Tuple[Any, ...],
+            kwargs: Mapping[str, Any],
+    ) -> Any:
+
+        # Diving into the instance to retrieve the Component name
+        component = instance.graph.nodes._nodes[args[0]]['instance']
+
+        if context_api.get_value(context_api._SUPPRESS_INSTRUMENTATION_KEY):
+            return wrapped(*args, **kwargs)
+
+        # Prepare invocation parameters by merging args and kwargs
+        invocation_parameters = {}
+        for arg in args:
+            if arg and isinstance(arg, dict):
+               invocation_parameters.update(arg)
+        invocation_parameters.update(kwargs)
+
+        span_name = component.__class__.__name__
+
+        with self._tracer.start_as_current_span(
+                span_name,
+                attributes={}
+        ) as span:
+            try:
+                response = wrapped(*args, **kwargs)
+            except Exception as exception:
+                span.set_status(trace_api.Status(trace_api.StatusCode.ERROR, str(exception)))
+                span.record_exception(exception)
+                raise
+            span.set_status(trace_api.StatusCode.OK)
+
+            attributes = dict(
```

Comment:
i would put the setting the `invocation_parameters` attributes above the `response = wrapped(*args, **kwargs)` 

consider the fact that if that line returns an exception, none of the invocation parameters would be set, which would be super debugging information exactly when the exception is happening!